### ISSUE TITLE
no-unused-vars flags global variables sometimes when it should not

### ISSUE THREAD

**Issue by ispringer** [eslint/eslint#1995]
I am getting numerous errors like the following:
```
  8:4  warning  formatters is defined but never used  no-unused-vars
```
when "formatters" is a global Object variable defined as follows:
```
var formatters = {
...
} ;
```
In my .eslintrc, I have this rule:
```
"no-unused-vars": [1, {"vars": "local", "args": "none"}], // default: [2, {"vars": "all", "args": "after-used"}]
```
Shouldn't the {"vars": "local"} option cause no error or warning to be output for the global "formatters" variable?

In addition, I have this in my .eslintrc:
```
"globals": {
    "formatters": true
}
```
Shouldn't this also cause the "formatters" variable to be ignored (even if I didn't have the rule above)?

I am using eslint 0.16.1.

**Comment by nzakas** [comment]
What else is in your .eslintrc file?

**Comment by ispringer** [comment]
Here is my .eslintrc file:
```
{
    "ecmaFeatures": {},
    "parser": "espree",
    "env": {
        "browser": true, // default: false
        "node": false,
        "amd": false,
        "mocha": false,
        "jasmine": false,
        "jquery": true // default: false
    },

    "globals": {
        "_": false,
        "angular": false,
        "kendo": false,
        "module": false,
        "require": false,
        "console": false,
        "formatters": true,
        "moment": false,
        "Highcharts": false,
        "chrome": false,
        "rangy": false,
        "tinycolor": false,
        "process": false,
        "ace": false,
        "RSVP": false,
        "SparkMD5": false,
        "HTMLParser": true
    },

    "rules": {
        "no-alert": 0, // default: 2
        "no-array-constructor": 2,
        "no-bitwise": 0,
        "no-caller": 2,
        "no-catch-shadow": 2,
        "no-comma-dangle": 2,
        "no-cond-assign": 2,
        "no-console": 2,
        "no-constant-condition": 2,
        "no-control-regex": 2,
        "no-debugger": 2,
        "no-delete-var": 2,
        "no-div-regex": 0,
        "no-dupe-keys": 2,
        "no-else-return": 0,
        "no-empty": 0, // default: 2
        "no-empty-class": 2,
        "no-empty-label": 2,
        "no-eq-null": 0,
        "no-eval": 2,
        "no-ex-assign": 2,
        "no-extend-native": 2,
        "no-extra-bind": 2,
        "no-extra-boolean-cast": 2,
        "no-extra-parens": 0,
        "no-extra-semi": 2,
        "no-extra-strict": 2,
        "no-fallthrough": 2,
        "no-floating-decimal": 0,
        "no-func-assign": 2,
        "no-implied-eval": 2,
        "no-inline-comments": 0,
        "no-inner-declarations": [2, "functions"],
        "no-invalid-regexp": 2,
        "no-irregular-whitespace": 2,
        "no-iterator": 2,
        "no-label-var": 2,
        "no-labels": 2,
        "no-lone-blocks": 2,
        "no-lonely-if": 0,
        "no-loop-func": 2,
        "no-mixed-requires": [0, false],
        "no-mixed-spaces-and-tabs": [2, false],
        "no-multi-spaces": 2,
        "no-multi-str": 2,
        "no-multiple-empty-lines": [0, {"max": 2}],
        "no-native-reassign": 2,
        "no-negated-in-lhs": 2,
        "no-nested-ternary": 0,
        "no-new": 2,
        "no-new-func": 2,
        "no-new-object": 2,
        "no-new-require": 0,
        "no-new-wrappers": 2,
        "no-obj-calls": 2,
        "no-octal": 2,
        "no-octal-escape": 2,
        "no-path-concat": 0,
        "no-plusplus": 0,
        "no-process-env": 0,
        "no-process-exit": 2,
        "no-proto": 2,
        "no-redeclare": 2,
        "no-regex-spaces": 2,
        "no-reserved-keys": 0,
        "no-restricted-modules": 0,
        "no-return-assign": 2,
        "no-script-url": 2,
        "no-self-compare": 0,
        "no-sequences": 2,
        "no-shadow": 2,
        "no-shadow-restricted-names": 2,
        "no-space-before-semi": 2,
        "no-spaced-func": 2,
        "no-sparse-arrays": 2,
        "no-sync": 0,
        "no-ternary": 0,
        "no-trailing-spaces": 2,
        "no-throw-literal": 0,
        "no-undef": 2,
        "no-undef-init": 2,
        "no-undefined": 0,
        "no-underscore-dangle": 0, // default: 2
        "no-unreachable": 2,
        "no-unused-expressions": 2,
        "no-unused-vars": [1, {"vars": "local", "args": "none"}], // default: [2, {"vars": "all", "args": "after-used"}]
        "no-use-before-define": 0, // default: 2
        "no-void": 0,
        "no-var": 0,
        "no-warning-comments": [0, { "terms": ["todo", "fixme", "xxx"], "location": "start" }],
        "no-with": 2,
        "no-wrap-func": 2,

        "block-scoped-var": 0,
        "brace-style": [0, "1tbs"],
        "camelcase": 0, // default: 2 // TODO: restore default
        "comma-spacing": 2,
        "comma-style": 0,
        "complexity": [0, 11],
        "consistent-return": 2,
        "consistent-this": [0, "that"],
        "curly": [2, "all"],
        "default-case": 0,
        "dot-notation": [2, { "allowKeywords": true }],
        "eol-last": 2,
        "eqeqeq": 0, // default: 2
        "func-names": 0,
        "func-style": [0, "declaration"],
        "generator-star": 0,
        "global-strict": [0, "never"], // default: [2, "never"]
        "guard-for-in": 0,
        "handle-callback-err": 0,
        "indent": 0,
        "key-spacing": [2, { "beforeColon": false, "afterColon": true }],
        "max-depth": [0, 4],
        "max-len": [0, 80, 4],
        "max-nested-callbacks": [0, 2],
        "max-params": [0, 3],
        "max-statements": [0, 10],
        "new-cap": 2,
        "new-parens": 2,
        "one-var": 0,
        "operator-assignment": [0, "always"],
        "padded-blocks": 0,
        "quote-props": 0,
        "quotes": [0, "double"], // default: [2, "double"]
        "radix": 0,
        "semi": 2,
        "sort-vars": 0,
        "space-after-function-name": [0, "never"],
        "space-after-keywords": [0, "always"],
        "space-before-blocks": [0, "always"],
        "space-before-function-parentheses": [0, "always"],
        "space-in-brackets": [0, "never"],
        "space-in-parens": [0, "never"],
        "space-infix-ops": 2,
        "space-return-throw-case": 2,
        "space-unary-ops": [2, { "words": true, "nonwords": false }],
        "spaced-line-comment": [0, "always"],
        "strict": 2,
        "use-isnan": 2,
        "valid-jsdoc": 0,
        "valid-typeof": 2,
        "vars-on-top": 0,
        "wrap-iife": 0,
        "wrap-regex": 0,
        "yoda": [2, "never"],

        // custom rules
        "no-lodash-aliases": 2
    }
}
```

**Comment by nzakas** [comment]
Thanks. So with the node environment, you variable is not being created in the global scope. Node has its own scope just under the global scope, so top-level variable and functions end up in that score instead of global.

**Comment by ispringer** [comment]
My application is browser-based, but I guess you mean that eslint is being run within node (I'm using grunt-eslint)? Is there some way for me to eliminate the warnings? Perhaps declaring my global variables like this:
```
window.formatters = {
...
};
```
rather than:
```
var formatters = {
...
};
```
?

**Comment by nzakas** [comment]
Oh, hmmm...for some reason I misread your config file as having `node: true`. Disregard my earlier comment and sorry for the confusion.

This does appear to be a bug.

**Comment by nzakas** [comment]
I've tried `var formatters = {}` both with `no-unused-vars` configured with just `locals` and defining `formatters` explicitly with `globals` and have been unable to reproduce the problem.

Do you have any other `.eslintrc` files in the project? Our configurations cascade, so they get combined.

Can you share what other code is in the file?

**Comment by ispringer** [comment]
No, there is only a single .eslintrc file in the top-level dir for our UI module. There is also a .jshintrc file in the same directory, but presumably that should be ignored. 

The "formatters" variable declaration lives in a file named formatters.js that's a few subdirectories deeper than the top-level dir. It looks like this:
```
/*
 * Copyright (C) 2010-2014 Evergage, Inc.
 * All rights reserved.
 */

'use strict';

var formatters = {
    shortDate: function(date) {
        return (date) ? kendo.toString(new Date(date), "MMM dd yyyy") : "";
    },

    shortDateKendo: "{0:MMM dd yyyy}",

    /* <snip> more function props similar to the above <snip> */
};
```

**Comment by nzakas** [comment]
Thanks, that's very helpful. As best I can tell from my quick debugging, `formatters` is being added to a function scope instead of the global scope. It looks like `ecmaFeatures.globalReturn` is somehow being toggled to `true`. Still unsure how that's happening, but investigating.

**Comment by nzakas** [comment]
Found it. Looks like we're not checking `true` or `false`, just the presence of the key. Easy fix.

**Comment by ispringer** [comment]
Great. Thank you!

**Comment by ispringer** [comment]
I just upgraded to eslint 0.16.2, and unfortunately I'm still seeing the warnings, e.g.:
```
src/app/utils/formatters.js
  8:4  warning  formatters is defined but never used  no-unused-vars
```

**Comment by nzakas** [comment]
That's frustrating, will have to investigate further.

**Comment by nzakas** [comment]
I'm unable to reproduce with 0.16.2. If I downgrade to 0.16.1, I see the error again. Then if I upgrade to 0.16.2, it disappears. This is what I'm seeing:

```
Nicholas@PATRIOT ~/projects/personal/jscheck ((v0.16.2))
$ git checkout v0.16.1
Previous HEAD position was 93a8925... 0.16.2
HEAD is now at 3bf1511... 0.16.1

Nicholas@PATRIOT ~/projects/personal/jscheck ((v0.16.1))
$ eslint ../tmp/foo.js

../tmp/foo.js
   3:4   warning  formatters is defined but never used           no-unused-vars
   6:1   error    Mixed spaces and tabs                          no-mixed-spaces
-and-tabs
   6:9   error    Multiple spaces found before '}'               no-multi-spaces

   8:1   error    Mixed spaces and tabs                          no-mixed-spaces
-and-tabs
   8:6   error    Multiple spaces found before 'shortDateKendo'  no-multi-spaces

   8:39  error    Trailing comma                                 no-comma-dangle

   8:39  error    Unexpected trailing comma                      comma-dangle
  10:1   error    Mixed spaces and tabs                          no-mixed-spaces
-and-tabs

✖ 8 problems (7 errors, 1 warning)


Nicholas@PATRIOT ~/projects/personal/jscheck ((v0.16.1))
$ git checkout v0.16.2
Previous HEAD position was 3bf1511... 0.16.1
HEAD is now at 93a8925... 0.16.2

Nicholas@PATRIOT ~/projects/personal/jscheck ((v0.16.2))
$ eslint ../tmp/foo.js

../tmp/foo.js
   6:1   error  Mixed spaces and tabs                          no-mixed-spaces-a
nd-tabs
   6:9   error  Multiple spaces found before '}'               no-multi-spaces
   8:1   error  Mixed spaces and tabs                          no-mixed-spaces-a
nd-tabs
   8:6   error  Multiple spaces found before 'shortDateKendo'  no-multi-spaces
   8:39  error  Trailing comma                                 no-comma-dangle
   8:39  error  Unexpected trailing comma                      comma-dangle
  10:1   error  Mixed spaces and tabs                          no-mixed-spaces-a
nd-tabs

✖ 7 problems (7 errors, 0 warnings)
```

Can you double-check that you've upgraded to 0.16.2 successfully?

**Comment by nzakas** [comment]
note: `jscheck` was the original project name before I found out it was taken. :) Just never changed my directory.

**Comment by ispringer** [comment]
I just figured out I was still running 0.16.1. grunt-eslint has a dependency on eslint "^0.16.1", which I had tried overriding to "0.16.2" in a few different ways, but apparently none of them worked. I ended up just manually copying the eslint 0.16.2 files over top of node_modules/grunt-eslint/node_modules/eslint/, in order to try running 0.16.2 on my project via grunt (where we specify which js files to include, etc.). Once I did this, the invalid warnings disappeared.

I'm still not sure why 0.16.2 wasn't getting pulled down and used automatically, since I'm pretty sure "^0.16.1" should match any 0.x.x version. It could be something to do with the local npm lazy mirror that my team uses not working as I expected.

 Sorry for the false alarm, and thanks again!
